### PR TITLE
feat: add leveldb/1.22

### PR DESCRIPTION
**Pull request by gocarlos**
Fixes https://github.com/conan-io/wishlist/issues/256 and fixes https://github.com/conan-io/wishlist/issues/25

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by gocarlos** [comment]
addresses also https://github.com/conan-io/wishlist/issues/256

File: recipes/leveldb/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    # def configure(self):
+
+    def requirements(self):
+        if self.options.with_snappy:
+            self.requires.add("snappy/1.1.8")
```
Comment (prince-chrismc) [inline recipes/leveldb/all/conanfile.py:42]:
```suggestion
            self.requires("snappy/1.1.8")
```

File: recipes/leveldb/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
+    # def configure(self):
+
```
Comment (prince-chrismc) [inline recipes/leveldb/all/conanfile.py:39]:
```suggestion
```

File: recipes/leveldb/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+from conans import CMake, ConanFile, tools
+from conans.errors import ConanInvalidConfiguration
+from conans.tools import Version
+
+
+class LevelDBCppConan(ConanFile):
+    name = "leveldb"
+    description = "LevelDB is a fast key-value storage library written at Google that provides an ordered mapping from string keys to string values."
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://gitlab.com/google/leveldb"
```
Comment (prince-chrismc) [inline recipes/leveldb/all/conanfile.py:11]:
```suggestion
    homepage = "https://github.com/google/leveldb"
```

File: recipes/leveldb/all/test_package/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+cmake_minimum_required(VERSION 2.8.11)
+project(test_package)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
+
+add_executable(${PROJECT_NAME} test_package.cpp)
+target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
+set_property(TARGET ${PROJECT_NAME} PROPERTY CXX_STANDARD 17)
```
Comment (prince-chrismc) [inline recipes/leveldb/all/test_package/CMakeLists.txt:9]:
```suggestion
```

**Comment by Croydon** [comment]
Please put `Fixes https://github.com/conan-io/wishlist/issues/256 and fixes https://github.com/conan-io/wishlist/issues/25` in the PR description, so that those issues get automatically closed when merged. Thanks 😄

File: recipes/leveldb/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        cmake = self._configure_cmake()
+        cmake.install()
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
+
+    def package_info(self):
+        self.cpp_info.libs = tools.collect_libs(self)
+        if not self.options.shared:
+            if self.settings.os == "Linux":
+                self.cpp_info.system_libs = ["pthread"]
```
Comment (uilianries) [inline recipes/leveldb/all/conanfile.py:71]:
```suggestion
        if self.settings.os == "Linux":
            self.cpp_info.system_libs = ["pthread"]
```

Keep it as usual

Comment (gocarlos) [reply recipes/leveldb/all/conanfile.py:71]: @madebr told me to do so?

Comment (gocarlos) [reply recipes/leveldb/all/conanfile.py:71]: Above

Comment (madebr) [reply recipes/leveldb/all/conanfile.py:71]: I checked and there is no mention of pthread in the headers.
So this should be just fine.

Comment (uilianries) [reply recipes/leveldb/all/conanfile.py:71]: ldd shows pthread, @gocarlos , keep it as usual

Comment (madebr) [reply recipes/leveldb/all/conanfile.py:71]: You need to run `objdump -p`.
See the `security` section of `ldd`:
```
   Security
       Be  aware  that  in  some circumstances (e.g., where the program specifies an ELF interpreter other than ld-
       linux.so), some versions of ldd may attempt to obtain the dependency information by attempting  to  directly
       execute the program, which may lead to the execution of whatever code is defined in the program's ELF inter‐
       preter, and perhaps to execution of the program itself.  (In glibc versions before 2.27,  the  upstream  ldd
       implementation did this for example, although most distributions provided a modified version that did not.)

       Thus,  you  should  never  employ  ldd on an untrusted executable, since this may result in the execution of
       arbitrary code.  A safer alternative when dealing with untrusted executables is:

           $ objdump -p /path/to/program | grep NEEDED

       Note, however, that this alternative shows only the direct dependencies of the executable, while  ldd  shows
       the entire dependency tree of the executable.

```

Comment (uilianries) [reply recipes/leveldb/all/conanfile.py:71]: good tip! that's why I use containers or vms haha

**Comment by gocarlos** [comment]
> --

?

File: recipes/leveldb/all/conanfile.py (hunk 5 above)
Comment (madebr) [reply recipes/leveldb/all/conanfile.py:71]: The last line is important: `objdump -p` shows only the direct dependencies of the executable, while  ldd  shows  the entire dependency tree of the executable

File: recipes/leveldb/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+from conans.errors import ConanInvalidConfiguration
+from conans.tools import Version
+
+
+class LevelDBCppConan(ConanFile):
+    name = "leveldb"
+    description = "LevelDB is a fast key-value storage library written at Google that provides an ordered mapping from string keys to string values."
+    url = "https://github.com/conan-io/conan-center-index"
+    homepage = "https://github.com/google/leveldb"
+    topics = ("conan", "leveldb", "db")
```
Comment (jjkoshy) [inline recipes/leveldb/all/conanfile.py:12]:
```suggestion
topics = ("conan", "leveldb", "google", "db")

File: recipes/leveldb/all/conanfile.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    license = "MIT"
```
Comment (jjkoshy) [inline recipes/leveldb/all/conanfile.py:13]:
```suggestion
license = ("BSD-3-Clause",)
```

File: recipes/leveldb/all/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    _cmake = None
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    def config_options(self):
+        if self.settings.os == "Windows":
+            del self.options.fPIC
+
```
Comment (jjkoshy) [inline recipes/leveldb/all/conanfile.py:37]:
```suggestion

    # FIXME: crc32, tcmalloc are also conditionally included in leveldb, but
    # there are no "official" conan packages yet; when those are available, we
    # can add similar with options for those
    
```

File: recipes/leveldb/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+    settings = "os", "arch", "compiler", "build_type"
+    options = {
+        "shared": [True, False], 
+        "with_snappy": [True, False],
+        "fPIC": [True, False]
+    }
+    default_options = {
+        "shared": False,
+        "fPIC": True,
+        "with_snappy": True,
```
Comment (jjkoshy) [inline recipes/leveldb/all/conanfile.py:25]: did you by any chance try `with_snappy` true and another snappy installed in your system? I can give this a try as well, but I think I ran into an issue with the wrong library (the system-installed one) getting picked. That's why I added patch sources portion in https://github.com/conan-io/conan-center-index/pull/1542/files#diff-224f12e422e56324bbccab54a890d9b3R57
although it has been a while since I worked on it so I may not be remembering details right. (I had written this a while ago in my repo: https://github.com/jjkoshy/conan-recipes/tree/master/conan-leveldb)

File: recipes/leveldb/all/test_package/test_package.cpp [hunk 10]
```diff
@@ -0,0 +1,20 @@
+#include "leveldb/db.h"
+#include <iostream>
+
+int main()
+{
+  leveldb::DB *db;
+  leveldb::Options options;
+  options.create_if_missing = true;
+  leveldb::Status status = leveldb::DB::Open(options, "/tmp/testdb", &db);
```
Comment (jjkoshy) [inline recipes/leveldb/all/test_package/test_package.cpp:9]: In #1542 I tried to avoid a platform-specific tmp path, but as long as CI is passing, then :man_shrugging:

File: recipes/leveldb/all/conanfile.py (hunk 9 above)
Comment (jjkoshy) [reply recipes/leveldb/all/conanfile.py:25]: So I think I remember this better now - basically, if a build specifies `with_snappy` false but the underlying machine has a pre-installed snappy in a standard library location, then libleveldb will incorporate references to that snappy library. i.e., if a consumer of the package wishes to use leveldb without snappy, there will be link errors. So that's why I think we should ideally patch leveldb's cmakelists.txt if `with_snappy` is false.

File: recipes/leveldb/all/test_package/test_package.cpp (hunk 10 above)
Comment (uilianries) [inline recipes/leveldb/all/test_package/test_package.cpp:9]:
```suggestion
  leveldb::Status status = leveldb::DB::Open(options, "testdb", &db);
```

What happen if you use locally?

File: recipes/leveldb/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+    def _configure_cmake(self):
+        if self._cmake:
+            return self._cmake
+        self._cmake = CMake(self)
+        self._cmake.definitions["LEVELDB_BUILD_TESTS"] = False
+        self._cmake.definitions["LEVELDB_BUILD_BENCHMARKS"] = False
+        self._cmake.configure()
+        return self._cmake
+
+    def build(self):
```
Comment (jjkoshy) [inline recipes/leveldb/all/conanfile.py:60]:
```suggestion
    def _patch_sources(self):
        if not self.options.with_snappy:
            tools.replace_in_file(
                    os.path.join(self._source_subfolder, "CMakeLists.txt"),
                    ('''check_library_exists(snappy snappy_compress '''
                        '''"" HAVE_SNAPPY)'''), "")

    def build(self):
        self._patch_sources()
```

File: recipes/leveldb/all/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        extracted_dir = self.name + "-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def _configure_cmake(self):
+        if self._cmake:
+            return self._cmake
+        self._cmake = CMake(self)
+        self._cmake.definitions["LEVELDB_BUILD_TESTS"] = False
+        self._cmake.definitions["LEVELDB_BUILD_BENCHMARKS"] = False
+        self._cmake.configure()
```
[the remark this answers is not in the document]
Comment (jjkoshy) [reply recipes/leveldb/all/conanfile.py:57]: @gocarlos below is a suggested change to avoid this issue.

This matters if you happen to use the recipe on a system that has a local libsnappy installed in which case, all the HAVE_SNAPPY sections in this file gets included in libleveldb: https://github.com/google/leveldb/blob/201f52201f5dd9701e7a8ceaa0ec4d344e69e022/port/port_stdcxx.h#L88-L99 causing the test_package build to fail linking:

```
[100%] Linking CXX executable bin/test_package
/usr/bin/ld: /home/jjkoshy/.conan/data/leveldb/1.22/_/_/package/b911f48570f9bb2902d9e83b2b9ebf9d376c8c56/lib/libleveldb.a(table_builder.cc.o): in function `leveldb::TableBuilder::WriteBlock(leveldb::BlockBuilder*, leveldb::BlockHandle*)':
table_builder.cc:(.text+0x629): undefined reference to `snappy::MaxCompressedLength(unsigned long)'
/usr/bin/ld: table_builder.cc:(.text+0x661): undefined reference to `snappy::RawCompress(char const*, unsigned long, char*, unsigned long*)'
/usr/bin/ld: /home/jjkoshy/.conan/data/leveldb/1.22/_/_/package/b911f48570f9bb2902d9e83b2b9ebf9d376c8c56/lib/libleveldb.a(format.cc.o): in function `leveldb::ReadBlock(leveldb::RandomAccessFile*, leveldb::ReadOptions const&, leveldb::BlockHandle const&, leveldb::BlockContents*)':
format.cc:(.text+0x48f): undefined reference to `snappy::GetUncompressedLength(char const*, unsigned long, unsigned long*)'
/usr/bin/ld: format.cc:(.text+0x4b3): undefined reference to `snappy::RawUncompress(char const*, unsigned long, char*)'
collect2: error: ld returned 1 exit status
```

In other words, the package's libleveldb depends on snappy regardless of the recipe's option to disable snappy.

The fix is simple - if have_snappy is set to false, then just patch CMakeLists.txt to remove the HAVE_SNAPPY portion.

File: recipes/leveldb/all/conanfile.py (hunk 11 above)
Comment (gocarlos) [reply recipes/leveldb/all/conanfile.py:67]: thanks

**Comment by jjkoshy** [comment]
Thanks for working on this @gocarlos
@uilianries there is an "empty" changes requested comment from you. If this looks good to you can we get this merged in?
